File: pact-jvm-server/src/main/scala/au/com/dius/pact/server/Create.scala
package au.com.dius.pact.server

import java.io.{File, IOException}
import java.net.ServerSocket
import au.com.dius.pact.consumer.model.{MockHttpsProviderConfig, MockProviderConfig}
import au.com.dius.pact.core.model._
import com.typesafe.scalalogging.StrictLogging
import org.apache.commons.lang3.RandomUtils

import java.security.KeyStore
import scala.collection.JavaConverters._

object Create extends StrictLogging {

  def create(state: String, path: List[String], requestBody: String, oldState: ServerState, config: Config): Result = {
    val pact = DefaultPactReader.INSTANCE.loadPact(requestBody).asInstanceOf[RequestResponsePact]

    val mockConfig : MockProviderConfig = {
      if (config.getKeystorePath.nonEmpty) {
        val ks = KeyStore.getInstance(new File(config.getKeystorePath), config.getKeystorePassword.toCharArray)
        new MockHttpsProviderConfig(
          config.getHost,
          config.getSslPort,
          PactSpecVersion.fromInt(config.getPactVersion),
          ks,
          "localhost",
          config.getKeystorePassword,
          config.getKeystorePassword
        )
      }
      else {
        new MockProviderConfig(config.getHost, randomPort(config.getPortLowerBound, config.getPortUpperBound),
          PactSpecVersion.fromInt(config.getPactVersion))
      }
    }
    val server = DefaultMockProvider.INSTANCE.apply(mockConfig)

    val port = server.getConfig.getPort
    val portEntry = port.toString -> server

    // Not very scala...
    val newState = (oldState.getState.asScala + portEntry) ++
      (for (
        pathValue <- path
      ) yield (pathValue -> server))

    val body = OptionalBody.body(("{\"port\": " + port + "}").getBytes)

    server.start(pact)

    new Result(new Response(201, (ResponseUtils.CrossSiteHeaders ++ Map("Content-Type" -> List("application/json").asJava)).asJava, body), new ServerState(newState.asJava))
  }

  def apply(request: Request, oldState: ServerState, config: Config): Result = {
    def errorJson = OptionalBody.body("{\"error\": \"please provide state param and path param and pact body\"}".getBytes)
    def clientError = new Result(new Response(400, ResponseUtils.CrossSiteHeaders.asJava, errorJson),
      oldState)

    logger.debug(s"path=${request.getPath}")
    logger.debug(s"query=${request.getQuery}")
    logger.debug(request.getBody.toString)

    val result = if (request.getQuery != null) {
      for {
        stateList <- CollectionUtils.javaLMapToScalaLMap(request.getQuery).get("state")
        state <- stateList.headOption
        paths <- CollectionUtils.javaLMapToScalaLMap(request.getQuery).get("path")
        body <- Option(request.getBody)
      } yield create(state, paths, body.valueAsString(), oldState, config)
    } else None

    result getOrElse clientError
  }

  def randomPort(lower: Int, upper: Int): Int = {
    var port: Integer = null
    var count = 0
    while (port == null && count < 20) {
      val randomPort = RandomUtils.nextInt(lower, upper)
      if (portAvailable(randomPort)) {
        port = randomPort
      }
      count += 1
    }

    if (port == null) {
      port = 0
    }

    port
  }

  private def portAvailable(p: Int): Boolean = {
    var socket: ServerSocket = null
    try {
      socket = new ServerSocket(p)
      true
    } catch {
      case _: IOException => false
    } finally {
      if (socket != null) {
        try {
          socket.close()
        } catch {
          case _: IOException => {}
        }
      }
    }
  }
}

File: pact-jvm-server/src/main/kotlin/au/com/dius/pact/server/Create.kt
package au.com.dius.pact.server

import au.com.dius.pact.consumer.model.MockHttpsProviderConfig
import au.com.dius.pact.consumer.model.MockProviderConfig
import au.com.dius.pact.core.model.DefaultPactReader
import au.com.dius.pact.core.model.OptionalBody
import au.com.dius.pact.core.model.PactSpecVersion
import au.com.dius.pact.core.model.Request
import au.com.dius.pact.core.model.Response
import au.com.dius.pact.core.support.isNotEmpty
import io.github.oshai.kotlinlogging.KotlinLogging
import org.apache.commons.lang3.RandomUtils
import java.io.File
import java.io.IOException
import java.net.ServerSocket
import java.security.KeyStore

private val logger = KotlinLogging.logger {}

object Create {

  private val CrossSiteHeaders = mapOf("Access-Control-Allow-Origin" to listOf("*"))

  @JvmStatic
  fun create(_state: String, path: List<String?>?, requestBody: String, oldState: ServerState, config: Config): Result {
    val pact = DefaultPactReader.loadPact(requestBody)

    val mockConfig : MockProviderConfig = if (config.keystorePath.isNotEmpty()) {
      val ks = KeyStore.getInstance(File(config.keystorePath), config.keystorePassword.toCharArray())
      MockHttpsProviderConfig(
        config.host,
        config.sslPort,
        PactSpecVersion.fromInt(config.pactVersion),
        ks,
        "localhost",
        config.keystorePassword,
        config.keystorePassword
      )
    }
    else {
      MockProviderConfig(config.host, randomPort(config.portLowerBound, config.portUpperBound),
        PactSpecVersion.fromInt(config.pactVersion))
    }
    val server = DefaultMockProvider.apply(mockConfig)

    val port = server.config.port
    val portEntry = port.toString() to server

    val newState = (oldState.state.entries.map { it.toPair() } + portEntry).toMutableSet()
    if (path != null) {
      for (p in path) {
        if (p != null) {
          newState += (p to server)
        }
      }
    }

    val body = OptionalBody.body(("{\"port\": $port}").toByteArray())

    server.start(pact)

    val headers = CrossSiteHeaders + mapOf("Content-Type" to listOf("application/json"))
    return Result(Response(201, headers.toMutableMap(), body), ServerState(newState.associate { it }))
  }

  @JvmStatic
  fun apply(request: Request, oldState: ServerState, config: Config): Result {
    logger.debug { "path=${request.path}" }
    logger.debug { "query=${request.query}" }
    logger.debug { "body=${request.body}" }

    if (request.query.isNotEmpty()) {
      val stateList = request.query["state"]
      if (stateList != null) {
        val state = stateList.first()
        if (state != null) {
          val paths = request.query["path"]
          if (request.body.isPresent()) {
            return create(state, paths, request.body.valueAsString(), oldState, config)
          }
        }
      }
    }

    val errorJson = OptionalBody.body("{\"error\": \"please provide state param and path param and pact body\"}".toByteArray())
    return Result(Response(400, CrossSiteHeaders.toMutableMap(), errorJson), oldState)
  }

  private fun randomPort(lower: Int, upper: Int): Int {
    var port: Int? = null
    var count = 0
    while (port == null && count < 20) {
      val randomPort = RandomUtils.nextInt(lower, upper)
      if (portAvailable(randomPort)) {
        port = randomPort
      }
      count += 1
    }

    if (port == null) {
      port = 0
    }

    return port
  }

  private fun portAvailable(p: Int): Boolean {
    var socket: ServerSocket? = null
    return try {
      socket = ServerSocket(p)
      true
    } catch (_: IOException) {
      false
    } finally {
      if (socket != null) {
        try {
          socket.close()
        } catch (_: IOException) {
        }
      }
    }
  }
}
